source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc

```diff
@@ -32,68 +38,241 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
   const std::string target = "cpu  ";
   if (!cpu_stats_file || buffer != target) {
     ENVOY_LOG_MISC(error, "Unexpected format in linux cpu stats file {}", cpu_stats_filename_);
-    return {false, 0, 0};
+    return {false, false, 0, 0, 0};
   }
 
   std::array<uint64_t, NUMBER_OF_CPU_TIMES_TO_PARSE> times;
   for (uint64_t time, i = 0; i < NUMBER_OF_CPU_TIMES_TO_PARSE; ++i) {
     cpu_stats_file >> time;
     if (!cpu_stats_file) {
       ENVOY_LOG_MISC(error, "Unexpected format in linux cpu stats file {}", cpu_stats_filename_);
-      return {false, 0, 0};
+      return {false, false, 0, 0, 0};
     }
     times[i] = time;
   }
 
   uint64_t work_time, total_time;
   work_time = times[0] + times[1] + times[2]; // user + nice + system
   total_time = work_time + times[3];          // idle
-  return {true, static_cast<double>(work_time), total_time};
+  return {true, false, static_cast<double>(work_time), total_time, 0};
+}
+
+// Container CPU Stats Reader
+LinuxContainerCpuStatsReader::ContainerStatsReaderPtr
+LinuxContainerCpuStatsReader::create(Filesystem::Instance& fs, TimeSource& time_source) {
+  // Check if host supports cgroup v2
+  if (CpuPaths::isV2(fs)) {
+    return std::make_unique<CgroupV2CpuStatsReader>(fs, time_source);
+  }
+
+  // Check if host supports cgroup v1
+  if (CpuPaths::isV1(fs)) {
+    return std::make_unique<CgroupV1CpuStatsReader>(fs, time_source);
+  }
+
+  throw EnvoyException("No supported cgroup CPU implementation found");
+}
+
+CgroupV1CpuStatsReader::CgroupV1CpuStatsReader(Filesystem::Instance& fs, TimeSource& time_source)
+    : LinuxContainerCpuStatsReader(fs, time_source), shares_path_(CpuPaths::V1::getSharesPath()),
+      usage_path_(CpuPaths::V1::getUsagePath()) {}
+
+CgroupV1CpuStatsReader::CgroupV1CpuStatsReader(Filesystem::Instance& fs, TimeSource& time_source,
+                                               const std::string& shares_path,
+                                               const std::string& usage_path)
+    : LinuxContainerCpuStatsReader(fs, time_source), shares_path_(shares_path),
+      usage_path_(usage_path) {}
+
+CpuTimes CgroupV1CpuStatsReader::getCpuTimes() {
+  // Read cpu.shares (cpu allocated)
+  auto shares_result = fs_.fileReadToEnd(shares_path_);
+  if (!shares_result.ok()) {
+    ENVOY_LOG(error, "Unable to read CPU shares file at {}", shares_path_);
+    return {false, false, 0, 0, 0};
+  }
+
+  // Read cpuacct.usage (cpu times)
+  auto usage_result = fs_.fileReadToEnd(usage_path_);
+  if (!usage_result.ok()) {
+    ENVOY_LOG(error, "Unable to read CPU usage file at {}", usage_path_);
+    return {false, false, 0, 0, 0};
+  }
+
+  TRY_ASSERT_MAIN_THREAD {
+    const double cpu_allocated_value = std::stod(shares_result.value());
+    const double cpu_times_value = std::stod(usage_result.value());
```

Comment:
can you use the absl numeric libs here: https://github.com/abseil/abseil-cpp/blob/5b1e199db4bc8f83f3d2dc1ced3dee7576ceedc2/absl/strings/numbers.h#L65-L88

this way we can avoid needing a try-catch and here as it was prior for cgroupv1